# Download Data

In [16]:
from sqlalchemy import create_engine
import pymysql
import pandas as pd

In [28]:
db_connection_str = 'mysql+pymysql://deepanalytics:Sqltask1234!@34.73.222.197/deepanalytics'
db_connection = create_engine(db_connection_str)
df = pd.read_sql('SELECT * FROM credit', con=db_connection)
### write to csv
df.to_csv('dataset_M2_Original.csv', index=False) #index=False
df.head()

MyUnknownColumn         X1      X2          X3        X4   X5     X6     X7  \
0              ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
1               1      20000  female  university         1   24      2      2   
2               2     120000  female  university         2   26     -1      2   
3               3      90000  female  university         2   34      0      0   
4               4      50000  female  university         1   37      0      0   

      X8     X9  ...        X15        X16        X17       X18       X19  \
0  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
1     -1     -1  ...          0          0          0         0       689   
2      0      0  ...       3272       3455       3261         0      1000   
3      0      0  ...      14331      14948      15549      1518      1500   
4      0      0  ...      28314      28959      29547      2000      2019   

        X20       X21       X22       X23                           Y  
0  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0         0                     default  
2      1000      1000         0      2000                     default  
3      1000      1000      1000      5000                 not default  
4      1200      1100      1069      1000                 not default  

[5 rows x 25 columns]

# EDA

In [76]:
# DS Bsics
import numpy as np
import pandas as pd
from pandas_profiling import ProfileReport
import matplotlib.pyplot as plt
import seaborn as sns

In [77]:
# import data
df = pd.read_csv('dataset_M2_Original.csv')
df.head()

MyUnknownColumn         X1      X2          X3        X4   X5     X6     X7  \
0              ID  LIMIT_BAL     SEX   EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
1               1      20000  female  university         1   24      2      2   
2               2     120000  female  university         2   26     -1      2   
3               3      90000  female  university         2   34      0      0   
4               4      50000  female  university         1   37      0      0   

      X8     X9  ...        X15        X16        X17       X18       X19  \
0  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
1     -1     -1  ...          0          0          0         0       689   
2      0      0  ...       3272       3455       3261         0      1000   
3      0      0  ...      14331      14948      15549      1518      1500   
4      0      0  ...      28314      28959      29547      2000      2019   

        X20       X21       X22       X23                           Y  
0  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0         0                     default  
2      1000      1000         0      2000                     default  
3      1000      1000      1000      5000                 not default  
4      1200      1100      1069      1000                 not default  

[5 rows x 25 columns]

Nice, let's take a look to the categorical variables. 

In [78]:
df[['X2', 'X3', 'X4']].describe()

X2          X3     X4
count    30204       30204  30204
unique       4           6      6
top     female  university      2
freq     18217       14107  16088

There is something strange in this DF. 4 values un the column X2 (SEX). Let's print the unique values it has:

In [79]:
df.X2.unique()

array(['SEX', 'female', 'male', 'X2'], dtype=object)

the words 'SEX' and 'X2' shouldn't be there.. Let's see whichs rows contains tose words, to perform that we can use .loc 

In [80]:
df.loc[df['X2']=='SEX']

MyUnknownColumn         X1   X2         X3        X4   X5     X6     X7  \
0                ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   
203              ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2   

        X8     X9  ...        X15        X16        X17       X18       X19  \
0    PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   
203  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2   

          X20       X21       X22       X23                           Y  
0    PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
203  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  

[2 rows x 25 columns]

well.. rows 0 and 203 contains strange things, that are, basically, no usefull data. 
We have to remove them, and we can do that in different ways. 

In [82]:
# in this case we're performing a copy of rows that NOT contain the word 'X2'
df = df[df.X2 != 'X2']
# and let's take a look to the output
df.X2.unique()

array(['SEX', 'female', 'male'], dtype=object)

worked nice, repeat with 'SEX'

In [83]:
df = df[df.X2 != 'SEX']
df.X2.unique()

array(['female', 'male'], dtype=object)

Nice now, at least X2, seems clean. let's repeat with X3 and X4

In [86]:
print(df.X3.unique())
print(df.X4.unique())

['university' 'graduate school' 'high school' 'other']
['1' '2' '3' '0']


There are no strange values and the description and amount of values matches the ones in the documentation. 

In [84]:
df[['X2', 'X3', 'X4']].describe()

X2          X3     X4
count    30201       30201  30201
unique       2           4      4
top     female  university      2
freq     18217       14107  16088